# Limpieza del dataset de bienes raíces

Este es un conjunto de datos (dataset) reales que fue descargado usando técnicas de web scraping. El archivo contiene registros de **Fotocasa**, el cual es uno de los sitios más populares de bienes raíces en España. Contiene miles de datos de casas reales publicadas en la web www.fotocasa.com.

El dataset fue descargado hace algunos años y en ningún caso se obtuvo beneficio económico de ello.

Tu objetivo es extraer tanta información como sea posible con el conocimiento que tienes hasta ahora de ciencia de datos.

### Observaciones respecto a la entrega

La tarea fue pensada en conjunto con la compañera de curso Leonor Bentancor.

In [2]:
import pandas as pd  # Importo librería pandas

#### Ejercicio 00

Lee el dataset data/real_estate.csv e intenta visualizar la tabla (★☆☆)

In [3]:
# Este archivo CSV contiene puntos y comas en lugar de comas como separadores
df = pd.read_csv('../data/real_estate.csv', sep=';')
df

Unnamed: 0  id_realEstates  isNew                    realEstate_name  \
0               1       153771986  False            ferrari 57 inmobiliaria   
1               2       153867863  False  tecnocasa fuenlabrada ferrocarril   
2               3       153430440  False                 look find boadilla   
3               4       152776331  False  tecnocasa fuenlabrada ferrocarril   
4               5       153180188  False            ferrari 57 inmobiliaria   
...           ...             ...    ...                                ...   
15330       15331       153901377  False                infocasa consulting   
15331       15332       150394373  False                inmobiliaria pulpon   
15332       15333       153901397  False             tecnocasa torrelodones   
15333       15334       152607440  False                inmobiliaria pulpon   
15334       15335       153901356  False                infocasa consulting   

       phone_realEstate                                       url_inmueble  \
0           912177526.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
1           916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
2           916350408.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
3           916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
4           912177526.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
...                 ...                                                ...   
15330       911360461.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15331       912788039.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15332       912780348.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15333       912788039.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15334       911360461.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   

       rooms  bathrooms  surface   price  ... level4Id level5Id level6Id  \
0        3.0        2.0    103.0  195000  ...        0        0        0   
1        3.0        1.0      NaN   89000  ...        0        0        0   
2        2.0        2.0     99.0  390000  ...        0        0        0   
3        3.0        1.0     86.0   89000  ...        0        0        0   
4        2.0        2.0    106.0  172000  ...        0        0        0   
...      ...        ...      ...     ...  ...      ...      ...      ...   
15330    2.0        1.0     96.0  259470  ...        0        0        0   
15331    3.0        1.0    150.0  165000  ...        0        0        0   
15332    4.0        2.0    175.0  495000  ...        0        0        0   
15333    3.0        2.0    101.0  195000  ...        0        0        0   
15334    3.0        2.0    152.0  765000  ...        0        0        0   

      level7Id level8Id accuracy          latitude          longitude zipCode  \
0            0        0        0  40,2948276786438  -3,44402412135624     NaN   
1            0        0        1          40,28674           -3,79351     NaN   
2            0        0        0  40,4115646786438  -3,90662252135624     NaN   
3            0        0        0  40,2853785786438  -3,79508142135624     NaN   
4            0        0        0  40,2998774864376  -3,45226301356237     NaN   
...        ...      ...      ...               ...                ...     ...   
15330        0        0        0          40,45416           -3,70286     NaN   
15331        0        0        0          40,36652           -3,48951     NaN   
15332        0        0        0          40,57444           -3,92124     NaN   
15333        0        0        0          40,36967           -3,48105     NaN   
15334        0        0        0          40,45773           -3,69068     NaN   

      customZone  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
15330        NaN  
15331        NaN  
15332        NaN  
15333        NaN  
15334        NaN  

[15335

In [4]:
df.info()  # Consulto información relevante del dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15335 entries, 0 to 15334
Data columns (total 37 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        15335 non-null  int64  
 1   id_realEstates    15335 non-null  int64  
 2   isNew             15335 non-null  bool   
 3   realEstate_name   15325 non-null  object 
 4   phone_realEstate  14541 non-null  float64
 5   url_inmueble      15335 non-null  object 
 6   rooms             14982 non-null  float64
 7   bathrooms         14990 non-null  float64
 8   surface           14085 non-null  float64
 9   price             15335 non-null  int64  
 10  date              15335 non-null  object 
 11  description       15193 non-null  object 
 12  address           15335 non-null  object 
 13  country           15335 non-null  object 
 14  level1            15335 non-null  object 
 15  level2            15335 non-null  object 
 16  level3            15335 non-null  object

## Trabajando con un DataFrame

#### Ejercicio 01

¿Cuál es la casa más cara del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada. Para visualizar el resultado utiliza un f-string. Por ejemplo:

```py
f'La casa más cara se encuentra en la dirección: {address} y su precio es {price} €'
```

In [5]:
precio_maximo = df['price'].max()  # Busco el máximo de la columna precio

casa_mas_cara = df[df['price'] == precio_maximo]  # Filtro el dataframe por las filas con precio máximo
address = casa_mas_cara['address'].values[0]  # Busco la primera fila del dataframe filtrado
price = casa_mas_cara['price'].values[0]

print(f'La casa más cara se encuentra en la dirección: {address} y su precio es {price} €')


La casa más cara se encuentra en la dirección: El Escorial y su precio es 8500000 €


#### Ejercicio 02

¿Cuál es la casa más barata del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada utilizando f-string

In [6]:
precio_minimo = df['price'].min()  # Realizo las mismas operaciones pero para el mínimo

casa_mas_barata = df[df['price'] == precio_minimo]
address_min = casa_mas_barata['address'].values[0]
price_min = casa_mas_barata['price'].values[0]

print(f'La casa más barata se encuentra en la dirección: {address_min} y su precio es {price_min} €')

La casa más barata se encuentra en la dirección: Parla y su precio es 0 €


#### Ejercicio 03

¿Cuál es la casa más grande del dataset? (★☆☆)

Imprime la dirección y el área de las casas seleccionadas utilizando f-string

In [7]:
tamanio_maximo = df['surface'].max()

casa_mas_grande = df[df['surface'] == tamanio_maximo]
address_max = casa_mas_grande['address'].values[0]
surface_max = casa_mas_grande['surface'].values[0]

print(f'La casa más grande se encuentra en la dirección: {address_max} y su superficie es de {surface_max}')

La casa más grande se encuentra en la dirección: Sevilla la Nueva y su superficie es de 249000.0


#### Ejercicio 04

¿Cuál es la casa más pequeña del dataset? (★☆☆)

In [8]:
tamanio_minimo = df['surface'].min()

casa_mas_pequenia = df[df['surface'] == tamanio_minimo]
address_min = casa_mas_pequenia['address'].values[0]
surface_min = casa_mas_pequenia['surface'].values[0]

print(f'La casa más pequeña se encuentra en la dirección: {address_min} y su superficie es de {surface_min}')

La casa más pequeña se encuentra en la dirección: Calle Amparo,  Madrid Capital y su superficie es de 15.0


#### Ejercicio 05.

¿El dataset contiene valores no admitidos (NAs)? (★☆☆)

- Muestra el nombre de las filas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.
- También muestra el nombre de las columnas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.

In [9]:
df.isna().any(axis=1)  # Busco filas que contengan valores Nas

0        True
1        True
2        True
3        True
4        True
         ... 
15330    True
15331    True
15332    True
15333    True
15334    True
Length: 15335, dtype: bool

In [10]:
df.isna().any(axis=0)  # Busco columnas que contengan valores Nas

Unnamed: 0          False
id_realEstates      False
isNew               False
realEstate_name      True
phone_realEstate     True
url_inmueble        False
rooms                True
bathrooms            True
surface              True
price               False
date                False
description          True
address             False
country             False
level1              False
level2              False
level3              False
level4               True
level5              False
level6               True
level7               True
level8               True
upperLevel          False
countryId           False
level1Id            False
level2Id            False
level3Id            False
level4Id            False
level5Id            False
level6Id            False
level7Id            False
level8Id            False
accuracy            False
latitude            False
longitude           False
zipCode              True
customZone           True
dtype: bool

#### Ejercicio 06.

Elimina los NAs del dataset, si aplica (★★☆)

Muestra las dimensiones del DataFrame original y del DataFrame después de las eliminaciones.

In [11]:
df_sin_nas = df.dropna()  # Elimino lo Nas y se lo asigno a un nuevo dataframe
df.shape  # Muestro dimensiones del dataframe original y en la próxima celda del dataframe luego de eliminar Nas



(15335, 37)

In [12]:
df_sin_nas.shape

(0, 37)

#### Ejercicio 07

¿Cuantas poblaciones (columna level5) contiene el dataset? (★☆☆)

- Muestra una lista con los nombres de las poblaciones
- Muestra el total de las mismas

In [15]:
poblaciones_unicas = df['level5'].nunique()  # Busco cantidad de poblaciones únicas y se lo asigno a una variable
df['level5'].unique()  # Genero un array con las poblaciones únicas

array(['Arganda del Rey', 'Fuenlabrada', 'Boadilla del Monte',
       'Las Rozas de Madrid', ' Madrid Capital', 'Villaviciosa de Odón',
       'Pinto', 'Valdemoro', 'Navalcarnero', 'Pozuelo de Alarcón',
       'Torrejón de Ardoz', 'Navalagamella', 'San Sebastián de los Reyes',
       'Rivas-vaciamadrid', 'Alpedrete', 'Móstoles',
       'San Fernando de Henares', 'Coslada', 'Becerril de la Sierra',
       'Alcalá de Henares', 'Chinchón', 'Parla', 'Alcorcón',
       'El Escorial', 'Leganés', 'Pedrezuela', 'Majadahonda',
       'Villanueva de la Cañada', 'Villanueva del Pardillo',
       'Torrelodones', 'Moralzarzal', 'Mejorada del Campo', 'Aranjuez',
       'Corpa', 'Getafe', 'Velilla de San Antonio', 'Sevilla la Nueva',
       'San Martín de la Vega', 'Villalbilla', 'Collado Villalba',
       'Alcobendas', 'El Molar (Madrid)', 'Moraleja de Enmedio', 'Algete',
       'Campo Real', 'Torrejón de la Calzada', 'Colmenar Viejo',
       'Valdemorillo', 'Fuente El Saz de Jarama', 'Tres Cantos',

In [16]:
print(f"El total de poblaciones únicas es: {poblaciones_unicas}")

El total de poblaciones únicas es: 168


#### Ejercicio 08

¿Cuál es la media de precios en la población (columna level5) de "Arroyomolinos (Madrid)"? (★★☆)

In [17]:
media_arroyomolinos = df.loc [df['level5'] == 'Arroyomolinos (Madrid)', 'price'].mean()  # Filtro el dataframe por la población indicada y hago promedio de su columna precio

print("Media de precios en Arroyomolinos (Madrid):", media_arroyomolinos)

Media de precios en Arroyomolinos (Madrid): 294541.59854014596


#### Ejercicio 09.

¿Los precios promedios de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

- Muestra ambos promedios
- Escribe en una celda markdown una conclusión sobre ellos

In [18]:
media_valdemorillo = df.loc [df['level5'] == 'Valdemorillo', 'price'].mean()
media_galapagar = df.loc [df['level5'] == 'Galapagar', 'price'].mean()

print("Promedio promedio de Valdemorillo:", media_valdemorillo)
print("Precio promedio de Galapagar:", media_galapagar)

Promedio promedio de Valdemorillo: 363860.2931034483
Precio promedio de Galapagar: 360063.20238095237


La media de precios de inmuebles entre ambas localidades son similares, sin embargo, Valdemorillo es más costosa.

#### Ejercicio 10

¿Los promedios de precio por metro cuadrado (precio/m2) de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

> Pista: Crea una nueva columna llamada `pps` (*price per square* o precio por metro cuadrado) y luego analiza los valores.

- Muestra ambos promedios de precio por metro cuadrado
- Escribe en una celda markdown una conclusión sobre ellos

In [19]:
df['pps'] = df['price'] / df['surface']  # Creo una columna nueva con el precio/m2

In [20]:
media_valdemorillo_pps = df.loc [df['level5'] == 'Valdemorillo', 'pps'].mean()
media_galapagar_pps = df.loc [df['level5'] == 'Galapagar', 'pps'].mean()

print("Promedio promedio por metro cuadrado de Valdemorillo:", media_valdemorillo_pps)
print("Precio promedio por metro cuadrado de Galapagar:", media_galapagar_pps)

Promedio promedio por metro cuadrado de Valdemorillo: 1317.9502109024986
Precio promedio por metro cuadrado de Galapagar: 1606.3240303094024


Si bien los inmuebles de Valdemorillo son más costos, el metro cuadrado en Galapar tiene un mayor valor, esto sugiere o que hay más propiedades a la venta en Valdemorillo o que las propiedades de Galapar son más pequeñas.

#### Ejercicio 11

¿Cuántas agencia de bienes raíces contiene el dataset? (★★☆)

- Muestra el valor obtenido.

In [21]:
agencias_unicas = df['realEstate_name'].nunique()  # Busco la cantidad de agencias únicas del dataframe
print(f"El dataset contiene {agencias_unicas} agencias de bienes raíces únicas")

El dataset contiene 1821 agencias de bienes raíces únicas


#### Ejercicio 12

¿Cuál es la población (columna level5) que contiene la mayor cantidad de casas?(★★☆)

- Muestra la población y el número de casas.

In [22]:
poblacion_mas_casas = df['level5'].value_counts().idxmax()  # Busco el id de la población que tiene más registros para obtener el nombre
cantidad_casas = df['level5'].value_counts().max()  # Busco la cantidad de registros de la población que tiene el máximo

print(f"La población que con tiene más casas es {poblacion_mas_casas} con {cantidad_casas} casas")

La población que con tiene más casas es  Madrid Capital con 6643 casas


---

## Trabajando con un subconjunto del DataFrame

#### Ejercicio 13

Ahora vamos a trabajar con el "cinturón sur" de Madrid.

Haz un subconjunto del DataFrame original que contenga las siguientes poblaciones (columna level5): "Fuenlabrada", "Leganés", "Getafe", "Alcorcón" (★★☆)

> Pista: Filtra el DataFrame original usando la columna `level5` y la función `isin`.

In [23]:
poblaciones_cinturon_sur = ["Fuenlabrada", "Leganés", "Getafe", "Alcorcón"]  # Genero una lista con las poblaciones a filtrar posteriormente
df_cinturon_sur = df[df['level5'].isin(poblaciones_cinturon_sur)]  # Creo un dataframe nuevo que contiene las poblaciones de la lista

#### Ejercicio 14

Calcula la media y la varianza de muestra para las siguientes variables: precio, habitaciones, superficie y baños (★★★)

> Debes usar el subset obtenido en la pregunta 13

- Crea y visualiza un diccionario con todos los valores

In [24]:
variables = ['price', 'rooms', 'surface', 'bathrooms']
df_cinturon_sur_estadisticas = df_cinturon_sur[variables].agg(['mean','var'])  # Busco la media y varianza para las variables que definí previamente
df_cinturon_sur_estadisticas

price     rooms      surface  bathrooms
mean  2.230945e+05  3.017738   111.752224   1.633222
var   1.492137e+10  0.718886  4263.051760   0.571797

#### Ejercicio 15

¿Cuál es la casa más cara de cada población del cinturón sur de Madríd? (★★☆)

> Debes usar el subset obtenido en la pregunta 13

- Genera un DataFrame con esta información
- Muestra tanto la dirección como el precio de la casa seleccionada de cada población.
- Genera conclusiones en una celda markdown

In [25]:
indices_casas_caras = df_cinturon_sur.groupby('level5')['price'].idxmax()  # Busco los índices de las casas más caras, agrupadas por población
df_casas_caras_cinturon_sur = df_cinturon_sur.loc[indices_casas_caras]  # Creo un subconjunto con las casas más caras
display(df_casas_caras_cinturon_sur[['level5', 'address', 'price']])  # Muestro el subconjunto

level5                                address    price
5585      Alcorcón                               Alcorcón   950000
11561  Fuenlabrada  Calle de Paulo Freire, 5, Fuenlabrada   490000
2881        Getafe                                 Getafe  1050000
10412      Leganés           Avenida Reina Sofía, Leganés   650000

El precio de las casas más costosas varía considerablemente según la población en donde está ubicada dentro del cinturón sur de Madrid.
Mientras que en Fuenlabrada el costo máximo es de € 490.000 en Getafe es de € 1.050.000, es decir, más de el doble.

#### Ejercicio 16

¿Qué puedes decir acerca del precio por metro cuadrado (precio/m2) entre los municipios de 'Getafe' y 'Alcorcón'? (★★☆)

> Debes usar el subset obtenido en la pregunta 13
>
> Pista: Crea una nueva columna llamada `pps` (price per square en inglés) y luego analiza los valores

In [103]:
print("Getafe:")
print(df_cinturon_sur[df_cinturon_sur['level5'] == 'Getafe']['pps'].describe())  # Genero datos estadísticos del pps para las poblaciones solicitadas y las muestro en pantalla para sacar conclusiones

print("\nAlcorcón:")
print(df_cinturon_sur[df_cinturon_sur['level5'] == 'Alcorcón']['pps'].describe())


Getafe:
count     241.000000
mean     2066.314949
std       741.872702
min         0.000000
25%      1684.285714
50%      1973.333333
75%      2628.787879
max      3827.160494
Name: pps, dtype: float64

Alcorcón:
count     173.000000
mean     2239.302480
std       539.951527
min       604.761905
25%      1904.081633
50%      2207.792208
75%      2472.727273
max      3698.159509
Name: pps, dtype: float64


Si bien el precio promedio por metro cuadrado de Getafe es menor (€ 2.066 vs € 2239), el 75% de los inmuebles de Alcorcón tiene menor pps que Getafe (€ 2.472 vs € 2.628), esto sugiere que en Getafe hay más casas "top" (entre el percentil 75 y el máximo)

## Conclusiones

#### Ejercicio 17

Fue un proyecto integrador que me pareció muy útil para aplicar los conocimientos del curso.
Es gratificante poder sacar conclusiones a partir de datos que se consiguen mediante web scraping, porque genera una sensación de que se pueden explotar distintas fuentes de datos que a priori uno no las tiene en el radar.
También considero que es importante profundizar en la práctica e incorporar distintas habilidades para conformar datasets de calidad que permitan tomar mejores decisiones a partir de datos curados.